crates/bevy_math/src/curve/mod.rs

```diff
@@ -0,0 +1,610 @@
+        Ok(ProductCurve {
+            domain,
+            first: self,
+            second: other,
+            _phantom: PhantomData,
+        })
+    }
+
+    /// Create a new [`Curve`] by composing this curve end-to-end with another, producing another curve
+    /// with outputs of the same type. The domain of the other curve is translated so that its start
+    /// coincides with where this curve ends. A [`ChainError`] is returned if this curve's domain
+    /// doesn't have a finite right endpoint or if `other`'s domain doesn't have a finite left endpoint.
+    fn chain<C>(self, other: C) -> Result<ChainCurve<T, Self, C>, ChainError>
+    where
+        Self: Sized,
+        C: Curve<T>,
+    {
+        if !self.domain().is_right_finite() {
+            return Err(ChainError::RightInfiniteFirst);
+        }
+        if !other.domain().is_left_finite() {
+            return Err(ChainError::LeftInfiniteSecond);
+        }
+        Ok(ChainCurve {
+            first: self,
+            second: other,
+            _phantom: PhantomData,
+        })
+    }
+
+    /// Borrow this curve rather than taking ownership of it. This is essentially an alias for a
+    /// prefix `&`; the point is that intermediate operations can be performed while retaining
+    /// access to the original curve.
+    ///
+    /// # Example
+    /// ```ignore
+    /// # use bevy_math::curve::*;
+    /// let my_curve = function_curve(interval(0.0, 1.0).unwrap(), |t| t * t + 1.0);
+    /// // Borrow `my_curve` long enough to resample a mapped version. Note that `map` takes
+    /// // ownership of its input.
+    /// let samples = my_curve.by_ref().map(|x| x * 2.0).resample_auto(100).unwrap();
+    /// // Do something else with `my_curve` since we retained ownership:
+    /// let new_curve = my_curve.reparametrize_linear(interval(-1.0, 1.0).unwrap()).unwrap();
+    /// ```
+    fn by_ref(&self) -> &Self
+    where
+        Self: Sized,
+    {
+        self
+    }
+
+    /// Flip this curve so that its tuple output is arranged the other way.
+    fn flip<U, V>(self) -> impl Curve<(V, U)>
+    where
+        Self: Sized + Curve<(U, V)>,
+    {
+        self.map(|(u, v)| (v, u))
+    }
+}
+
+impl<T, C, D> Curve<T> for D
+where
+    C: Curve<T> + ?Sized,
+    D: Deref<Target = C>,
+{
+    fn domain(&self) -> Interval {
+        <C as Curve<T>>::domain(self)
+    }
+
+    fn sample(&self, t: f32) -> T {
+        <C as Curve<T>>::sample(self, t)
+    }
+}
+
+/// An error indicating that an end-to-end composition couldn't be performed because of
+/// malformed inputs.
+#[derive(Debug, Error)]
+#[error("Could not compose these curves together")]
+pub enum ChainError {
+    /// The right endpoint of the first curve was infinite.
+    #[error("The first curve has an infinite right endpoint")]
+    RightInfiniteFirst,
+
+    /// The left endpoint of the second curve was infinite.
+    #[error("The second curve has an infinite left endpoint")]
+    LeftInfiniteSecond,
```

Comment:
I don't know if those are known names, but I feel like
`FirstRightInfinite` and `SecondLeftInfinite` are more readable.

Reply:
FirstEndInfinite gets my vote.

Reply:
Well then you should revoke your start as left comment, or use `FirstRightInfinite` here